components/layout/query.rs

```diff
@@ -613,6 +613,15 @@ fn ensure_node_data_initialized<N: LayoutNode>(node: &N) {
     let mut cur = Some(node.clone());
     while let Some(current) = cur {
         if current.borrow_layout_data().is_some() {
+            if cfg!(debug_assertions) {
+                cur = current.parent_node();
+                while let Some(current) = cur {
+                    assert!(current.borrow_layout_data().is_some(),
+                            "Broken invariant: Child has data without parent \
+                             having it");
+                    cur = current.parent_node();
+                }
+            }
```

Comment:
That looks pretty bad, given it's already in a loop that calls `parent_node`. As for the invariant, I'm not sure I should be the one to review this.